src/validation/tests/assignment_validation_tests.rs

```diff
@@ -1217,3 +1217,94 @@ fn void_assignment_validation() {
+                referenceToFooInitializedLiteral    : REFERENCE TO DINT := 5;
+                referenceToFooInitializedArray      : REFERENCE TO ARRAY[1..5] OF DINT;
+            END_VAR
+
+            refToFoo        REF= foo;
+            referenceToFoo  REF= foo;
+
+            // Invalid
+            1 REF= foo;
+            foo REF= foo;
+            referenceToFoo REF= 0;
+            referenceToFoo REF= referenceToFoo;
+        END_FUNCTION
+        ",
+    );
+
+    assert_snapshot!(diagnostics, @r###"
+    error[E099]: REFERENCE TO variables can not reference other variables
+       ┌─ <internal>:18:55
+       │
+    18 │                 referenceToFooInitializedVariable   : REFERENCE TO foo;
+       │                                                       ^^^^^^^^^^^^^^^^ REFERENCE TO variables can not reference other variables
+
+    error[E099]: REFERENCE TO variables can not reference other variables
+       ┌─ <internal>:19:55
+       │
+    19 │                 referenceToFooGlobalVariable        : REFERENCE TO fooGlobal;
+       │                                                       ^^^^^^^^^^^^^^^^^^^^^^ REFERENCE TO variables can not reference other variables
+
+    error[E099]: REFERENCE TO variables can not reference arrays, pointers or bits
+       ┌─ <internal>:21:17
+       │
+     4 │             AliasedArray : ARRAY[1..5] OF DINT;
+       │             ------------ see also
+       ·
+    21 │                 referenceToFooAliasedArrayType      : REFERENCE TO AliasedArray;
+       │                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ REFERENCE TO variables can not reference arrays, pointers or bits
+
+    error[E099]: Initializations of REFERENCE TO variables are disallowed
+       ┌─ <internal>:22:76
+       │
+    22 │                 referenceToFooInitializedLiteral    : REFERENCE TO DINT := 5;
+       │                                                                            ^ Initializations of REFERENCE TO variables are disallowed
+
+    error[E099]: REFERENCE TO variables can not reference arrays, pointers or bits
+       ┌─ <internal>:23:17
+       │
+    23 │                 referenceToFooInitializedArray      : REFERENCE TO ARRAY[1..5] OF DINT;
+       │                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^        -------------------------------- see also
+       │                 │                                      
+       │                 REFERENCE TO variables can not reference arrays, pointers or bits
+
+    error[E098]: Invalid assignment, expected a reference
+       ┌─ <internal>:30:13
+       │
+    30 │             1 REF= foo;
+       │             ^ Invalid assignment, expected a reference
+
+    error[E098]: Invalid assignment, expected a reference
+       ┌─ <internal>:32:33
+       │
+    32 │             referenceToFoo REF= 0;
+       │                                 ^ Invalid assignment, expected a reference
+
+    "###);
+}
```

Comment:
can you add a few further assignment tests for more complex types/mismatches, i.e. STRING to WSTRING/WCHAR etc. here?